Logging (#33)
meilisearch/specifications: text/0033-logging.md

*Elastic App Search's Log Retention settings view.*

> It is also possible to do that from a Log Settings API endpoint.

> API Logs is aimed for trace requests and responses at engine level while Analytics Logs trace query, clicks and counts.

### IV. Explanation

#### Current Logging behaviour of MeiliSearch (0.20)

MeiliSearch uses `env_logger` to allow setting the output level of logs from the `RUST_LOG` environment variable. `env_logger` permits to set a specific log level per module if needed. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).

#### Logging behaviour for Milli (0.21)

We have decided to keep the use of `env_logger` for Milli/Transplant. However, we are going to make changes to make the logging more consistent and more versatile.

##### Log Levels

| Level | Description                                                                                                              |
|-------|--------------------------------------------------------------------------------------------------------------------------|
| ERROR | Everything related to an error occuring.                                                                                         |

> | WARN  | Everything related to a warning. TBD                                                                                     |

Reviewer: I need to know what is the `TBD` before opening an issue 🙂

Reviewer: Or maybe we don't add warnings for the v0.21.0? I don't know if the spec concerns only the v0.21.0 or the future releases as well

> | INFO  | Info should be the default level. It displays high level of informations like start/end of process and HTTP calls.                                            |

Reviewer: We might come back to you to know if a process deserves to be logged or not since there is not accurate list at the moment 🙂

Log level can be: `debug`, `info`, `warn`, `error`, `fatal`, or `unknown`.

#### Elastic Entreprise Search

Elasticsearch's SaaS platform, like Algolia, offers a quite similar endpoint to browse the logs generated by the system.

However, it is possible to deactivate the storage of logs, modify the retention period but also to filter the logs with more details.

![image](https://www.elastic.co/guide/en/app-search/current/images/guides/log-settings-controls.png)
*Elastic App Search's Log Retention settings view.*

> It is also possible to do that from a Log Settings API endpoint.

> API Logs is aimed for trace requests and responses at engine level while Analytics Logs trace query, clicks and counts.

### IV. Explanation

#### Current Logging behaviour of MeiliSearch (0.20)

MeiliSearch uses `env_logger` to allow setting the output level of logs from the `RUST_LOG` environment variable. `env_logger` permits to set a specific log level per module if needed. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).

#### Logging behaviour for Milli (0.21)

> We have decided to keep the use of `env_logger` for Milli/Transplant. However, we are going to make changes to make the logging more consistent and more versatile.

Reviewer: Same as MeiliSearch v0.20.0, should we let the usage of `RUST_LOG` env variable? This is not mentioned in this part the user can access different level of log via this variable

Author: Yes!

### V. Impact on Documentation

The documentation only mention the logging behavior for the `development` env on the `MEILI_ENV` part.

We should explain how to specify the log level using `RUST_LOG` environment variable and display the Log Levels table as information in a dedicated section.

### VI. Impact on SDKs
N/A

## 2. Technical Aspects

MeiliSearch is not consistent about logging methods. It have occurences of `println` and `eprintln` in the codebase and occurences of `error`, `warn`, `info`, `debug`, `trace` methods. It also calls `log::warn` or `log::error`.

Milli and Transplant should be careful by keeping a consistent way to log informations.

## 3. Future Possibilities

- Store logs on filesystem (give us future possibilites of rolling strategy). We will keep an eye on https://roadmap.meilisearch.com/c/81-specify-log-path, Github issues and, Slack Community messages. Keep in mind that it is possible to send logs to files using `syslog` or `systemd` journalctl.
- Develop API endpoint to search the logged events and configure logging policy within the system (SaaS feature in mind).

## 4. Planned Changes

### 0.21

#### Core

> - Use a consistent method to log

Reviewer: ```suggestion
- Use a consistent method to log (relative to internal implementation)
```

> - Log output should start with the mandatory log format part.

Reviewer: Should we implement the HTTP call format you describe? I would say yes so that we are iso with MeiliSearch v0.20.0 :)

Author: Yes!

- Title: Logging
- Start Date: 2021-04-21
- Specification PR: [#33](https://github.com/meilisearch/specifications/pull/33)

> - MeiliSearch Tracking-issues:

Reviewer: ```suggestion
- MeiliSearch Tracking-issues: [transplant/#193](https://github.com/meilisearch/transplant/issues/193)
```

We should explain how to specify the logging level using the `RUST_LOG` environment variable and display the logging level table as information in a dedicated section.

### VI. Impact on SDKs
N/A

## 2. Technical Aspects

MeiliSearch is not consistent on logging methods. There are occurrences of `println` and `eprintln` in the codebase and occurrences of `error`, `warn`, `info`, `debug`, `trace` methods. It also calls `log::warn` or `log::error`.

Milli and Transplant must be careful in keeping a consistent way of logging information.

## 3. Future Possibilities

- Store logs on filesystem (give us future possibilites of rolling strategy). We will keep an eye on https://roadmap.meilisearch.com/c/81-specify-log-path, Github issues and, Slack Community messages. Keep in mind that it is possible to send logs to files using `syslog` or `systemd` journalctl.
- Develop an API endpoint to search for logged events and configure the logging policy for the instance (SaaS feature in mind).

## 4. Planned Changes

### 0.21

#### Core
- Use a consistent method to log (relative to internal implementation)

> - Log output should start with the mandatory log format part.

Reviewer: ```suggestion
- The HTTP logs should be logged as described in this spec
- The users should be able to choose the level of debug by filling the `RUST_LOG` environment variable
```

#### Elastic Entreprise Search

Elasticsearch's SaaS platform, like Algolia, offers a fairly similar endpoint for browsing system-generated logs.

However, it is possible to disable log storage, change the retention period and also filter the logs with more details.

![image](https://www.elastic.co/guide/en/app-search/current/images/guides/log-settings-controls.png)
*Elastic App Search's Log Retention settings view.*

> It is also possible to do that from a Log Settings API endpoint.

> API Logs track requests and responses at the engine level, while Analytics Logs track requests, clicks and counts.

### IV. Explanation

#### Current Logging behaviour of MeiliSearch (0.20)

MeiliSearch uses `env_logger` to allow to define the log output level from the `RUST_LOG` environment variable. `env_logger` allows to define a specific log level per module if necessary. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).

#### Logging behaviour for Milli (0.21)

We have decided to keep the use of `env_logger` for Milli/Transplant. However, we will make some changes to make the logging more consistent and versatile.

> We will expose a `LOG_LEVEL` flag that corresponds to `RUST_LOG` in the binary to improve its meaning and clarity to users.

Reviewer: `LOG_LEVEL` flag? Do you mean the env variable only or also a flag option like `--log-level`?

Author: 🙏 Thanks for the catch @curquiza. I've updated the specification to mean env variable.

> We will expose a `LOG_LEVEL` environment variable that corresponds to `RUST_LOG` in the binary to improve its meaning and clarity to users.

Reviewer 2: It is called `LOG_LEVEL` and not `MEILI_LOG_LEVEL`? All other environment variables are prefixed by `MEILI_`.

Author: Thanks, @qdequele! Nice catch!

> #### Current Logging behaviour of MeiliSearch (0.20)
> 
> MeiliSearch uses `env_logger` to allow to define the log output level from the `RUST_LOG` environment variable. `env_logger` allows to define a specific log level per module if necessary. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).
> 
> #### Logging behaviour for Milli (0.21)
> 
> We have decided to keep the use of `env_logger` for Milli/Transplant. However, we will make some changes to make the logging more consistent and versatile.
> 
> We will expose a `MEILI_LOG_LEVEL` environment variable that corresponds to `RUST_LOG` in the binary to improve its meaning and clarity to users. A CLI option `--log-level` will also be added to follow our configuration convention.

Reviewer 3: ```suggestion
#### Current Logging behaviour of MeiliSearch (0.20)

MeiliSearch uses the rust lib `env_logger` to output logs based on the log level fetched from the `RUST_LOG` environment variable. `env_logger` allows to define a specific log level per module if necessary. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).

#### Logging behaviour for Milli (0.21)

We have decided to keep using the rust lib `env_logger` for Milli/Transplant. However, we will make some changes to make the logging more consistent and versatile.

Instead of using `RUST_LOG`, we create a `MEILI_LOG_LEVEL` environment variable that corresponds to `RUST_LOG` in the binary to improve its meaning and clarity to users. A CLI option `--log-level` will also be added to follow our configuration convention.
```

##### Log Levels

| Level | Description                                                                                                       |
|-------|-------------------------------------------------------------------------------------------------------------------|
| ERROR | Everything related to an error. A non blocking error has occured.                                                 |
| WARN  | Used to warn about an exceptional non-blocking event occurring.                                                   |
| INFO  | Default Log Level. It displays high level informations of events occuring in the search engine.                   |
| DEBUG | Used for debugging and development purposes.  More verbose than INFO.                                             |

> | TRACE | Display everything happening at engine level. Not used at the moment.                                             |

Reviewer 3: Did we manage to find a solution to implement the TRACE level? As what I remember we could not have this level

Author: Hello @bidoubiwa. 

Sorry to answer lately! The trace level is used internally to debug complex issues. Although we can now use it internally, we will not document it to our users at first.

Reviewer 3: Any reason why it will not be documented for users?

Author: Yes, debug will be documented! People developing features with us in rust will be informed about this log level by us if they need an elaborate trace to tackle complex issues. We didn't want to document it to not overwhelm the users who won't really be concerned by it directly. The level trace is specifically for rust developers who will contribute to the engine and it will not be mandatory in the vast majority of cases.

Reviewer 3: If the user is overwhelmed he could go back to debug level. I'm not sure not documenting it is that impactful. We also  not documented served feature which is not a good thing 😕. Also if it is documented we can warn user on the overwhelming aspect of trace and we recommend using debug and only switching if debug is not good enough.

Author: Thanks @bidoubiwa! You are right about that documentation point. It's better to document it and let people see that when they possibly need it. We can document the `trace` level and providing some explanation that it is useful for rust developers to deal with complex problems that need to dive at the machine level. E.g. for expert only. I will update the specification.

Reviewer 4: More often than not, trace and debug hide issue. Debug and trace assume a knowledge of the codebase, and will be more confusing than helping for the user.

Author: Let's make the choice to have exhaustive documentation on log levels and see how it impacts our users negatively or positively. We can expect to have a really small impact on this decision.

- First line of request (Example: GET /test HTTP/1.1)
- Response status code
- Size of response in bytes, including HTTP headers
- User-Agent
- Time taken to serve the request, in seconds to 6 decimal places

> At DEBUG log level, the search endpoint must log the request body and the response body.

### V. Impact on Documentation

The documentation only mentions the logging behavior for the `development` env on the `MEILI_ENV` part.

We should explain how to specify the logging level using the `MEILI_LOG_LEVEL` environment variable and display the logging level table as information in a dedicated section. It should also mention the usage of the cli flag `--log-level`.

### VI. Impact on SDKs
N/A

## 2. Technical Aspects
N/A

## 3. Future Possibilities

- Store logs on filesystem (give us future possibilites of rolling strategy). We will keep an eye on https://roadmap.meilisearch.com/c/81-specify-log-path, Github issues and, Slack Community messages. Keep in mind that it is possible to send logs to files using `syslog` or `systemd` journalctl.
- Develop an API endpoint to search for logged events and configure the logging policy for the instance (SaaS feature in mind).

> - Add syntaxic sugar helper flag like `-v, -vv, -vvv` that can be translated to a `MEILI_LOG_LEVEL` value.
> 

Reviewer 5: ```suggestion
- Add syntactic sugar helper flag like `-v, -vv, -vvv` that can be translated to a `MEILI_LOG_LEVEL` value.
- Use the rust clone of `log4j2`, [`log4rs`](https://docs.rs/log4rs/1.0.0/log4rs/) instead of `env_logger` to provide the same kind of options that Elasticsearch propose.
```

From what I understand, from a user perspective, Elasticsearch has the best logging system, so I guess we should consider using `log4rs` to be as good as them.

Author: Indeed, this is something that could be added to allow users to have a more versatile logging policy, I imagine it will become very useful when we have high availability and logs will have to be collected remotely. The rolling log method could be very useful at that time.

- Time when the request was started to process (in rfc3339 format)
- Log levels are  `ERROR`, `WARN`, `INFO`, `DEBUG`, `TRACE`.
- The module part gives information about the module that records the log.

###### HTTP Call

Transplant uses `actix_web::middleware::logger` to record information about the API endpoints that receive calls.

Given
```172.17.0.1:57220 "POST /indexes/indexUID/documents HTTP/1.1" 202 14 "-" "PostmanRuntime/7.26.10" 0.023529```

- Peer IP address (or IP address of reverse proxy if used)
- First line of request (Example: GET /test HTTP/1.1)
- Response status code
- Size of response in bytes, including HTTP headers
- User-Agent
- Time taken to serve the request, in seconds to 6 decimal places

> At DEBUG log level, the search endpoint must log the request body and the response body.

### V. Impact on Documentation

The documentation only mentions the logging behavior for the `development` env on the `MEILI_ENV` part.

We should explain how to specify the logging level using the `LOG_LEVEL` environment variable and display the logging level table as information in a dedicated section.

### VI. Impact on SDKs
N/A

## 2. Technical Aspects

> MeiliSearch is not consistent on logging methods. There are occurrences of `println` and `eprintln` in the codebase and occurrences of `error`, `warn`, `info`, `debug`, `trace` methods. It also calls `log::warn` or `log::error`.

Author: Thanks for the precision @MarinPostma! I will update the spec according to that!
